[SPARK-7969] [SQL] Added a DataFrame.drop function that accepts a Column reference.

dusenberrymw opened the pull request:
Added a `DataFrame.drop` function that accepts a `Column` reference rather than a `String`, and added associated unit tests.  Basically iterates through the `DataFrame` to find a column with an expression that is equivalent to that of the `Column` argument supplied to the function.

dusenberrymw committed:
Added a DataFrame.drop function that accepts a Column reference rather than a String, and added associated unit tests.  Basically iterates through the DataFrame to find a column with an expression that is equivalent to one supplied to the function.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33983 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33983/consoleFull) for   PR 6585 at commit [`986129c`](https://github.com/apache/spark/commit/986129c4701b82b3cd0f83e235856f1fb29ce2d2).

SparkQA commented:
[Test build #33983 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33983/consoleFull) for   PR 6585 at commit [`986129c`](https://github.com/apache/spark/commit/986129c4701b82b3cd0f83e235856f1fb29ce2d2).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

ogirardot commented:
@rxin @marmbrus can you please review this PR ?

ogirardot commented:
Can you add a test with a join please ?

rxin commented:
cc @cloud-fan can you review this? 

@dusenberrymw can you make sure Python also works with this?

dusenberrymw committed:
Adding Python support for Dataframe.drop with a Column reference.

dusenberrymw commented:
@rxin Sure thing, just added Python support for the `drop` with a `Column` reference.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34036 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34036/consoleFull) for   PR 6585 at commit [`4b8bbe8`](https://github.com/apache/spark/commit/4b8bbe897ac1a884949836e897c2aea636c831a1).

SparkQA commented:
[Test build #34036 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34036/consoleFull) for   PR 6585 at commit [`4b8bbe8`](https://github.com/apache/spark/commit/4b8bbe897ac1a884949836e897c2aea636c831a1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala:
```diff
@@ -1083,6 +1083,21 @@ class DataFrame private[sql](
+   */
+  def drop(col: Column): DataFrame = {
+    val colsAfterDrop = schema.filter { field =>
+      !this.col(field.name).expr.semanticEquals(col.expr)
```
This is not right. If this data frame is a result of join, then it may have 2 columns with same name, and the `this.col(field.name)` will case a ambiguous-reference exception. You should use `df.logicalPlan.output` instead of `df.schema`.

dusenberrymw replied:
@cloud-fan Thanks for the advice, let me iterate on this a bit!

dusenberrymw committed:
Updating DataFrame.drop with column reference function to use logicalPlan.output to prevent ambiguities resulting from columns with the same name. Also added associated unit tests for joined datasets with duplicate column names.

dusenberrymw committed:
Adding more Python doctests for the df.drop with column reference function to test joined datasets that have columns with the same name.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34087 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34087/consoleFull) for   PR 6585 at commit [`e583888`](https://github.com/apache/spark/commit/e5838885d092b618b44dda88aff0ee198e1f01a8).

dusenberrymw commented:
Okay, I've updated the `DataFrame.drop` with `Column` references function to use `logicalPlan.output` instead of `schema` to avoid ambiguities for DataFrames with columns that have the same name.  I also added unit tests in Scala and Python specifically for testing this.

Now, if two DataFrames that share column names are joined together resulting in a DataFrame "joinedDf" with duplicate column names, the correct and appropriate column can be removed from the joined DataFrame by passing in the associated `Column` from the original DataFrame to `joinedDf.drop`.

@ogirardot @cloud-fan @rxin

SparkQA commented:
[Test build #34087 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34087/consoleFull) for   PR 6585 at commit [`e583888`](https://github.com/apache/spark/commit/e5838885d092b618b44dda88aff0ee198e1f01a8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -334,6 +334,50 @@ class DataFrameSuite extends QueryTest {
+    val df = joinedDf.drop(col)
+    checkAnswer(
+      df,
+      joinedDf.collect().map { case Row(id: Int, name: String, age: Int,
```
new line before `case`.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:
```diff
@@ -334,6 +334,50 @@ class DataFrameSuite extends QueryTest {
+                                        idToDrop: Int, salary: Double) =>
+        Row(id, name, age, salary)
+      }.toSeq)
+    assert(df.schema.map(_.name) === Seq("id", "name", "age", "salary"))
```
Maybe we should add `asset(df("id") == person("id"))`

cloud-fan commented:
LGTM exception some minor issues, thanks for it!

rxin commented on sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala:
```diff
@@ -1083,6 +1083,22 @@ class DataFrame private[sql](
+   * This is a no-op if the DataFrame doesn't have a column
+   * with an equivalent expression.
+   * @group dfops
+   * @since 1.4.0
```
note: since 1.4.1

dusenberrymw committed:
Minor code formatting change.

dusenberrymw committed:
Adding an additional assert statement to the 'drop column after join' unit test in order to make sure the correct column was indeed left over.

dusenberrymw committed:
Updating the @since tag of the drop(Column) function doc to reflect version 1.4.1 instead of 1.4.0.

dusenberrymw commented:
Thanks, @cloud-fan & @rxin!  I pushed the few minor changes, so please let me know if there is anything else we should change.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #34188 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34188/consoleFull) for   PR 6585 at commit [`514727a`](https://github.com/apache/spark/commit/514727aa621b812d4608a19a2a327e37393a85a7).

SparkQA commented:
[Test build #34188 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34188/consoleFull) for   PR 6585 at commit [`514727a`](https://github.com/apache/spark/commit/514727aa621b812d4608a19a2a327e37393a85a7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

rxin commented:
Thanks. I've merged this! It will be part of Spark 1.4.1 and 1.5.0.

dusenberrymw commented:
Great, thanks!
